### ISSUE TITLE
Notebook open failure is cached until reload

### ISSUE THREAD

**Issue by connor4312** [microsoft/vscode#125550]
1. Take the fsprovider sample: https://github.com/microsoft/vscode-extension-samples/tree/master/fsprovider-sample
2. Modify a line to write `memfs:/file.github-issues`, and modify line 11 to read:
    ```
        context.subscriptions.push(vscode.workspace.registerFileSystemProvider('memfs', memFs, { isCaseSensitive: true, isReadonly: false }));
    ```
3. Run the extension, execute `MemFS: Setup Workspace` and `MemFS: Create File`
4. You can open other files in the workspace ✔️ 
5. Try opening the github issues notebook, get this error 🐛 

![](https://memes.peet.io/img/21-06-1eef2d2e-af27-4b98-9f09-a2e659ac5f69.png)

Unlike other files, the `readFile` method on the FileSystemProvider is not called for a notebook.

**Comment by bpasero** [comment]
@connor4312 I cannot reproduce and the error rather seems to suggest that the file was not created initially?

Here is what I see:

![recording](https://user-images.githubusercontent.com/900690/120881707-73698a80-c5d3-11eb-8edb-5a92c4122a98.gif)

**Comment by connor4312** [comment]
Hm, I no longer can reproduce the error, although the notebook is always empty for me, even if I copy in the file contents of my-work.github-issues for example

**Comment by bpasero** [comment]
Yeah I can reproduce that too. From my end it seems the data is loaded properly and then passed to the notebook model, not sure why it would return empty:

![image](https://user-images.githubusercontent.com/900690/120882056-9f860b00-c5d5-11eb-97f3-503fc077d59e.png)

**Comment by jrieken** [comment]
The notebook works just fine for me using this snippet. Make sure to test this with well encoded JSON etc because the GH issue notebook will not print any error when the data is invalid. 

```ts
memFs.writeFile(vscode.Uri.parse(`memfs:/file.github-issues`), Buffer.from(JSON.stringify([{ kind: 1, value: '_hello_', language: 'markdown' }])), { create: true, overwrite: true });
```

**Comment by jrieken** [comment]
However, I can reproduce the file not found error and that's an ugly one

```
ERR Error: Unable to read file 'memfs:/file.github-issues' (Error: Unable to resolve non-existing file 'memfs:/file.github-issues')
    at FileService.doReadFileStream (fileService.js:425)
    at processTicksAndRejections (internal/process/task_queues.js:93)
    at async StoredFileWorkingCopy.resolveFromFile (storedFileWorkingCopy.js:273)
    at async StoredFileWorkingCopyManager.resolve (storedFileWorkingCopyManager.js:289)
    at async SimpleNotebookEditorModel.load (notebookEditorModel.js:424)
    at async NotebookModelReferenceCollection.createReferencedObject (notebookEditorModelResolverServiceImpl.js:65)
    at async NotebookModelResolverServiceImpl.resolve (notebookEditorModelResolverServiceImpl.js:146)
    at async NotebookEditorInput.resolve (notebookEditorInput.js:146)
    at async NotebookEditor.setInput (notebookEditor.js:143)
    at async EditorControl.doSetInput (editorControl.js:156)
    at async EditorControl.openEditor (editorControl.js:65)
    at async file:/Users/jrieken/Code/vscode/out/vs/workbench/browser/parts/editor/editorGroupView.js:752
```

Steps: 

* Add notebook to the mem-fs sample
* Run sample, create MemFS, create files, open notebook
* With the notebook open, reload window
* The notebook fails to be restored (because create files hasn't been run yet), you see the error above
* Run create files
* Re-try to open the notebook, the error-promise is cached and a proper retry is never attempted

**Comment by jrieken** [comment]
Closed in release/1.57. For main @joaomoreno is looking into a `AsyncReferenceCollection`
